mousestyles/ultradian/__init__.py

```diff
@@ -0,0 +1,360 @@
+    Aggregate the interval data based on n-minute time
+    intervals, return a time series.
+
+    Parameters
+    ---------------
+    strain: int
+        nonnegative integer indicating the strain number
+    mouse: int
+        nonnegative integer indicating the mouse number
+    feature: {"AS", "F", "IS", "M_AS", "M_IS", "W", "AS_prob"}
+    bin_width: number of minutes of time interval for data aggregation
+
+    Returns
+    ----------
+    ts: pandas.tseries
+        a pandas time series of length 12(day)*24(hour)*60(minute)/n
+    """
+    # load data
+    intervals = data.load_intervals(feature)
+    mouse_data = intervals.loc[
+        (intervals['strain'] == strain) & (intervals['mouse'] == mouse)]
+
+    # build data frame
+    days = sorted(np.unique(mouse_data['day']))
+    bin_count = int(24 * 60 / bin_width)
+    time_behaviour = np.repeat(0.0, bin_count * len(days))
+    bin_length = bin_width * 60
+
+    for j in days:
+        df = mouse_data.loc[mouse_data['day'] == j]
+        start_end = data.load_start_time_end_time(strain, mouse, j)
+        start = np.asarray(df['start']) - start_end[0]
+        end = np.asarray(df['stop']) - start_end[0]
+
+        for i in range(len(start)):
+            start_time = start[i]
+            end_time = end[i]
+            start_index = int(start_time / (bin_width * 60))
+            end_index = int(end_time / (bin_width * 60))
+            if start_index == end_index:
+                time_behaviour[start_index + j *
+                               bin_count] += end_time - start_time
+            elif end_index - start_index == 1:
+                time_behaviour[
+                    start_index + j *
+                    bin_count] += bin_length * end_index - start_time
+                time_behaviour[end_index + j *
+                               bin_count] += end_time % bin_length
+            else:
+                time_behaviour[
+                    start_index + j *
+                    bin_count] += bin_length * (start_index + 1) - start_time
+                time_behaviour[end_index + j *
+                               bin_count] += end_time % bin_length
+                time_behaviour[start_index + j * bin_count +
+                               1:end_index + j * bin_count] += bin_length
+
+    if feature == 'F' or feature == 'W':
+        all_feature = data.load_all_features()
+        group = all_feature[
+            ["strain", "mouse", "day", "hour", "Food", "Water"]].groupby(
+            ["strain", "mouse", "day"]).sum()
+        group = group.reset_index()
+        group['day'] -= 5
+        mouse_data = group.loc[(group['strain'] == strain) &
+                               (group['mouse'] == mouse)]
+        for i in mouse_data['day'].astype('int'):
+            if feature == 'F':
+                food_amount = float(mouse_data['Food'][mouse_data['day'] == i])
+                time_behaviour[
+                    (bin_count * i):(bin_count * (i + 1) - 1)] /= sum(
```

Comment:
Do we really need the -1 here? Because to subset the first n element, you just use [0, n] in python